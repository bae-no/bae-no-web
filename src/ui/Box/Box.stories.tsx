import { ComponentStory, ComponentMeta } from "@storybook/react";

import { colors } from "../tokens/color";
import { layout } from "../tokens/layout";
import { space } from "../tokens/space";

import Box from "./Box";

export default {
  argTypes: {
    alignItems: {
      control: "select",
      options: ["stretch", "flex-start", "center", "flex-end"],
    },
    backgroundColor: { control: "select", options: Object.keys(colors) },
    bg: { control: "select", options: Object.keys(colors) },
    borderColor: { control: "select", options: Object.keys(colors) },
    borderRadius: { control: "select", options: Object.keys(layout) },
    boxSizing: { control: "select", options: ["border-box", "content-box"] },
    br: { control: "select", options: Object.keys(layout) },
    color: { control: "select", options: Object.keys(colors) },
    cursor: { control: "select", options: ["pointer"] },
    display: {
      control: "select",
      options: ["none", "flex", "block", "inline", "grid"],
    },
    flex: { control: "select", options: [0, 0.5, 1] },
    flexDirection: {
      control: "select",
      options: ["row", "column", "row-reverse", "column-reverse"],
    },
    gap: { control: "select", options: Object.keys(space) },
    height: { control: "select", options: Object.keys(layout) },
    justifyContent: {
      control: "select",
      options: [
        "stretch",
        "flex-start",
        "center",
        "flex-end",
        "space-around",
        "space-between",
      ],
    },
    m: { control: "select", options: Object.keys(space) },
    margin: { control: "select", options: Object.keys(space) },
    marginBottom: { control: "select", options: Object.keys(space) },
    marginLeft: { control: "select", options: Object.keys(space) },
    marginRight: { control: "select", options: Object.keys(space) },
    marginTop: { control: "select", options: Object.keys(space) },
    maxHeight: { control: "select", options: Object.keys(layout) },
    maxSize: { control: "select", options: Object.keys(layout) },
    maxWidth: { control: "select", options: Object.keys(layout) },
    minHeight: { control: "select", options: Object.keys(layout) },
    minSize: { control: "select", options: Object.keys(layout) },
    minWidth: { control: "select", options: Object.keys(layout) },
    mx: { control: "select", options: Object.keys(space) },
    my: { control: "select", options: Object.keys(space) },
    p: { control: "select", options: Object.keys(space) },
    padding: { control: "select", options: Object.keys(space) },
    paddingBottom: { control: "select", options: Object.keys(space) },
    paddingLeft: { control: "select", options: Object.keys(space) },
    paddingRight: { control: "select", options: Object.keys(space) },
    paddingTop: { control: "select", options: Object.keys(space) },
    position: {
      control: "select",
      options: ["absolute", "relative", "fixed", "sticky"],
    },
    px: { control: "select", options: Object.keys(space) },
    py: { control: "select", options: Object.keys(space) },
    size: { control: "select", options: Object.keys(layout) },
    width: { control: "select", options: Object.keys(layout) },
    zIndex: { control: "select", options: [-1, 0, 1, 2, 3, 4, 5] },
  },
  component: Box,
  title: "UI/Box",
} as ComponentMeta<typeof Box>;

const Template: ComponentStory<typeof Box> = (args) => <Box {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: "Default Box",
};
