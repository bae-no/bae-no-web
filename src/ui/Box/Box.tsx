import {
  AllHTMLAttributes,
  ComponentType,
  createElement,
  ElementType,
  forwardRef,
} from "react";

import clsx, { ClassValue } from "clsx";

import { ComponentBaseProps } from "src/types";

import { sprinkles, Sprinkles } from "../sprinkles.css";
import { parseProps } from "../utils";

export interface BoxProps
  extends ComponentBaseProps,
    Omit<
      AllHTMLAttributes<HTMLElement>,
      | "className"
      | "content"
      | "translate"
      | "color"
      | "width"
      | "height"
      | "cursor"
      | "size"
      | "as"
    >,
    Sprinkles {
  as?: ElementType | ComponentType;
  className?: ClassValue;
  replace?: boolean; // FIXME: next 13 migration 중 임시처리
}

const Box = forwardRef(
  (
    {
      children,
      display = "flex",
      direction = "column",
      as = "div",
      className = "",
      ...props
    }: BoxProps,
    ref,
  ) => {
    const [atomProps, nativeProps] = parseProps(props);
    const styles = clsx(
      sprinkles({
        ...atomProps,
        direction,
        display,
      }),
      className,
    );

    return createElement(
      as,
      {
        className: styles,
        ...nativeProps,
        ref,
      },
      children,
    );
  },
);

export default Box;
