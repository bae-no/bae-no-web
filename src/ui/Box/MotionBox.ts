import { AllHTMLAttributes, ComponentType } from "react";

import { m } from "framer-motion";

import { ExcludeObject } from "src/types/ExcludeObject";

import Box, { BoxProps } from "./Box";

export const MotionBox = m(
  Box as ComponentType<
    ExcludeObject<
      BoxProps,
      Omit<
        AllHTMLAttributes<HTMLElement>,
        "children" | "height" | "width" | "as"
      >
    >
  >,
);
