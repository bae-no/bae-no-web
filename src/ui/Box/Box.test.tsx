import { fireEvent, render, screen } from "@testing-library/react";
import Box from "./Box";

describe("UI Box Component", () => {
  it("match snapshot", () => {
    const { container } = render(<Box flexDirection="row" p="sm" />);
    expect(container).toMatchSnapshot();
  });

  it("render text", () => {
    render(<Box>test text</Box>);
    expect(screen.getByText("test text")).toBeInTheDocument();
  });

  it("triggered click event", () => {
    const mockClick = jest.fn();
    render(<Box onClick={mockClick}>test text</Box>);
    fireEvent.click(screen.getByText("test text"));

    expect(mockClick).toHaveBeenCalled();
  });
});
