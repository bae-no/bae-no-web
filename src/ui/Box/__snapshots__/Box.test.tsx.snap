// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Box Component match snapshot 1`] = `
<div>
  <div
    class="sprinkles_padding_small_mobile__1vna7gb1r sprinkles_display_flex_mobile__1vna7gb3 sprinkles_flexDirection_row_mobile__1vna7gbf reset_base__4op0mk0"
  />
</div>
`;
